File: Java/URISC2.java
// A ISS for the URSIC microprocessor
import java.awt.*;
import java.awt.event.*;
import java.io.File;  // Import the File class
import java.io.FileNotFoundException;  // Import this class to handle errors
import java.util.Scanner; // Import the Scanner class to read text files

class urisc2 extends Frame {
    Button stepButton, resetButton;
    Label clkLabel, pcLabel, iwLabel;
    Label r0Label, r1Label, r2Label, r3Label, r4Label, r15Label;
    int [] prog = new int[4096];
    int [] reg = {5,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1};
    public urisc2 ()
    {   int k = 0, i=1, N=0;
        String fileName = "io.hex";
        try {
            File myObj = new File(fileName);
            Scanner myReader = new Scanner(myObj);
            while ((myReader.hasNextLine()) && (i>0)) {
                String data = myReader.nextLine();
                i = Integer.parseInt(data, 16);
                System.out.println(k + ") "+data + " = " + i);
                prog[k++] = i;
            }
            N=k;
            myReader.close();
        } catch (FileNotFoundException e) {
            System.out.println("File " + fileName + " not found!");
            e.printStackTrace();
        }
        setLayout(new GridLayout(0,2)); // 3 columns; infinity rows
        setTitle("ISS by UMB");
        add(new Label ("Name")); add(new Label ("Value"));
        // row 1: CLK and Value
        add(new Label ("CLK"));
        clkLabel = new Label("0");  add(clkLabel);
        // row 2: PC and value
        add(new Label ("PC"));
        pcLabel = new Label("0"); add(pcLabel);
        // row 3: IW and value
        add(new Label ("IW"));
        String result = String.format("%X", prog[0] );
        iwLabel = new Label(result); add(iwLabel);
        // row 4: R[0] and value
        add(new Label ("R[0]"));
        String r0 = String.format("%d", reg[0] );
        r0Label = new Label(r0); add(r0Label);
        // row 5: R[1] and value
        add(new Label ("R[1]"));
        String r1 = String.format("%d", reg[1] );
        r1Label = new Label(r1); add(r1Label);
        // row 6: R[2] and value
        add(new Label ("R[2]"));
        String r2 = String.format("%d", reg[2] );
        r2Label = new Label(r2); add(r2Label);
        //add(new Label ("Value"));
        // row 7: R[3] and value
        add(new Label ("R[3]"));
        String r3 = String.format("%d", reg[3] );
        r3Label = new Label(r3); add(r3Label);
        // row 8: R[4] and value
        add(new Label ("R[4]"));
        String r4 = String.format("%d", reg[4] );
        r4Label = new Label(r2); add(r4Label);
        // row 9: R[15] and value
        add(new Label ("R[15]"));
        String r15 = String.format("%d", reg[15] );
        r15Label = new Label(r15); add(r15Label);
        // Last row: Buttons for STEP and RESET
        stepButton = new Button ("STEP");
        stepButton.setBackground(new Color(0,128,0)); //
        stepButton.setForeground(Color.WHITE);
        add (stepButton);
        stepButton.addActionListener(new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                int clk = Integer.parseInt(clkLabel.getText());
                int pc_now = Integer.parseInt(pcLabel.getText());
                int pc_new;
                int dst = (prog[pc_now] &0xF000) >>12;
                int src = (prog[pc_now] & 0x0F00) >> 8;
                int mode = (prog[pc_now] & 0x0080) >> 7;
                int addr = (prog[pc_now] & 0x007F);
                int result = reg[dst] - reg[src];
                if (dst > 0)
                    reg[dst] = result;
                if (result >= 0) {
                    pc_new = pc_now + 1;
                } else if (mode == 1) {
                    pc_new = addr;
                } else {
                    pc_new = pc_now + addr;
                }
                // iw = prog[pc_new];
                //dst, src, mode, addr = decode(iw)
                result = reg[dst] - reg[src];
                String str = String.format("%d", ++clk); clkLabel.setText(str);
                str = String.format("%d", pc_new); pcLabel.setText(str);
                str = String.format("%X", prog[pc_new]); iwLabel.setText(str);
                str = String.format("%d", reg[0]); r0Label.setText(str);
                str = String.format("%d", reg[1]); r1Label.setText(str);
                str = String.format("%d", reg[2]); r2Label.setText(str);
                str = String.format("%d", reg[3]); r3Label.setText(str);
                str = String.format("%d", reg[4]); r4Label.setText(str);
                str = String.format("%d", reg[15]); r15Label.setText(str);
            }
        });
        resetButton = new Button ("RESET");
        resetButton.setBackground(Color.RED);
        resetButton.setForeground(Color.WHITE);
        add (resetButton);
        resetButton.addActionListener(new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                int i = Integer.parseInt (clkLabel.getText());
                String result = String.format("%d", 0);
                clkLabel.setText(result);
                pcLabel.setText(result);
                result = String.format("%X", prog[0]);
                iwLabel.setText(result);
                for (int k=1; k<=15; k++) reg[k] = -1;
                reg[0]= 5;
                String str = "5"; r0Label.setText(str);
                str = "-1"; r1Label.setText(str); r2Label.setText(str);
                r3Label.setText(str); r4Label.setText(str);
                r15Label.setText(str);
            }
        });
        addWindowListener(new WindowAdapter(){
            public void windowClosing(WindowEvent e) {
                dispose();
            }
        });

        setSize (250, 300);
        setVisible (true);
    }
    public static void main (String args[])
    {
        new urisc2();
    }
}

File: Python/URISC2.py
# A ISS for the URSIC microprocessor
import sys
import tkinter as tkinter

### Decode
def decode(iw):
    dst  = int(iw[0], 16)
    src  = int(iw[1], 16)
    imm  = int(iw[2:4], 16)
    mode = (imm & 0x0080) >> 7
    if mode==1: 
        addr = imm & 0x007F
    else:
        addr = imm & 0x003F
        sign = (imm & 0x0040) >> 6
        if sign==1:
            addr -= 64
    return dst, src, mode, addr

class App:
    def __init__(self, master):
        global prog
        global reg
        frame = tkinter.Frame(master)
        frame.pack()

        tkinter.Label(frame, text = 'Name').grid(row = 0, column = 0)
        tkinter.Label(frame, text = 'Value').grid(row = 0, column = 1)
        
        tkinter.Label(frame, text='CLK ').grid(row = 1, column = 0)
        self.clk = tkinter.IntVar(); self.clk.set(0)
        tkinter.Label(frame, textvariable=self.clk).grid(row=1, column=1)

        tkinter.Label(frame, text='PC ').grid(row=2, column=0)
        self.pc = tkinter.IntVar(); self.pc.set(0)
        tkinter.Label(frame, textvariable=self.pc).grid(row=2, column=1)

        tkinter.Label(frame, text='IW ').grid(row=3, column=0)
        self.iw = tkinter.StringVar();  self.iw.set(prog[0][0:4])
        tkinter.Label(frame, textvariable=self.iw).grid(row=3, column=1)
        
        tkinter.Label(frame, text='R[0] ').grid(row=4, column=0)
        self.r0 = tkinter.IntVar(); self.r0.set(reg[0])
        tkinter.Label(frame, textvariable=self.r0).grid(row=4, column=1)

        tkinter.Label(frame, text='R[1] ').grid(row=5, column=0)
        self.r1 = tkinter.IntVar(); self.r1.set(reg[1])
        tkinter.Label(frame, textvariable=self.r1).grid(row=5, column=1)
        
        tkinter.Label(frame, text='R[2] ').grid(row=6, column=0)
        self.r2 = tkinter.IntVar(); self.r2.set(reg[2])
        tkinter.Label(frame, textvariable=self.r2).grid(row=6, column=1)
        
        tkinter.Label(frame, text='R[3] ').grid(row=7, column=0)
        self.r3 = tkinter.IntVar(); self.r3.set(reg[3])
        tkinter.Label(frame, textvariable=self.r3).grid(row=7, column=1)

        tkinter.Label(frame, text='R[4] ').grid(row=8, column=0)
        self.r4 = tkinter.IntVar(); self.r4.set(reg[4])
        tkinter.Label(frame, textvariable=self.r4).grid(row=8, column=1)

        tkinter.Label(frame, text='R[15]').grid(row=9, column=0)
        self.r15 = tkinter.IntVar(); self.r15.set(reg[15])
        tkinter.Label(frame, textvariable=self.r15).grid(row=9, column=1)
        
        step = tkinter.Button(frame, text='STEP', command=self.step, \
                                               bg="green", fg="white")
        step.grid(row=10, column=0) 
        reset = tkinter.Button(frame, text='RESET', command=self.reset, \
                                                     bg="red", fg="white")
        reset.grid(row=10, column=1)

    def step(self):
        global prog
        global reg
    ### Fetch
        clock = self.clk.get()
        pc_now = self.pc.get()
        iw = prog[pc_now]
        dst, src, mode, addr = decode(iw)
    ### Execute
        result = reg[dst] - reg[src]
        if dst > 0:
            reg[dst] = result
        if result >= 0:
            pc_new = pc_now + 1
        elif mode == 1:
            pc_new = addr
        else:
            pc_new = pc_now + addr
        iw = prog[pc_new]
        dst, src, mode, addr = decode(iw)
        result = reg[dst] - reg[src]
        clock = clock + 1
        self.clk.set(clock);   self.pc.set(pc_new)
        self.iw.set(iw[0:4]);  self.r0.set(reg[0])
        self.r1.set(reg[1]);   self.r2.set(reg[2])
        self.r3.set(reg[3]);   self.r4.set(reg[4])
        self.r15.set(reg[15])

    def reset(self):
        global prog
        global reg 
        self.pc.set(0)
        self.clk.set(0)
        self.iw.set(prog[0][0:4])
        for k in range(16):
            reg[k]=-1
        reg[0]=in_port 
        self.r0.set(reg[0]);    self.r1.set(reg[1])
        self.r2.set(reg[2]);    self.r3.set(reg[3])
        self.r4.set(reg[4]);    self.r15.set(reg[15]) 

fin = open("io.hex", "r")
prog = fin.readlines()
fin.close()

in_port = 5
reg=[in_port]
for k in range(15):
    reg.append(-1)

root = tkinter.Tk()
root.wm_title('ISS by UMB')
root.geometry("300x300")
app = App(root)
root.mainloop()
